modules/programs/gh.nix

```
@@ -126,7 +126,25 @@ in {
     home.packages = [ cfg.package ];
 
     xdg.configFile."gh/config.yml".source =
-      yamlFormat.generate "gh-config.yml" cfg.settings;
+      yamlFormat.generate "gh-config.yml" ({ version = "1"; } // cfg.settings);
+
+    # Version 2.40.0+ of gh needs to migrate account formats, this needs to
+    # happen before the version = 1 is placed in the configuration file. Running
+    # `gh help` is sufficient to perform the migration. If the migration already
+    # has occurred, then this is a no-op.
+    #
+    # See https://github.com/nix-community/home-manager/issues/4744 for details.
+    home.activation.migrateGhAccounts =
+      hm.dag.entryBetween [ "linkGeneration" ] [ "writeBoundary" ] ''
+        if [[ -e "${config.xdg.configHome}/gh/config.yml" ]]; then
+          TMP_DIR=$(mktemp -d)
+          trap "rm --force --recursive $TMP_DIR" EXIT
+          cp "${config.xdg.configHome}/gh/hosts.yml" $TMP_DIR/
+          export GH_CONFIG_DIR=$TMP_DIR
+          $DRY_RUN_CMD ${getExe cfg.package} help 2>&1 > $DRY_RUN_NULL
+          cp $TMP_DIR/hosts.yml "${config.xdg.configHome}/gh/hosts.yml"
```

I think the `TMP_DIR` variable name needs to be more unique, to avoid issues with duplicate variable names in the activation script. Similarly, it is a good to unset variables that are no longer used.

So I would suggest something like this:
```suggestion
          GH_ACCOUNT_MIGRATION_DIR=$(mktemp -d)
          trap "rm --force --recursive $GH_ACCOUNT_MIGRATION_DIR" EXIT
          cp "${config.xdg.configHome}/gh/hosts.yml" $GH_ACCOUNT_MIGRATION_DIR/
          export GH_CONFIG_DIR=$GH_ACCOUNT_MIGRATION_DIR
          $DRY_RUN_CMD ${getExe cfg.package} help 2>&1 > $DRY_RUN_NULL
          cp $GH_ACCOUNT_MIGRATION_DIR/hosts.yml "${config.xdg.configHome}/gh/hosts.yml"
          unset GH_CONFIG_DIR
```

---

In fact, perhaps it would be a good idea to use a subshell for this. Then there wouldn't be an issue with the risk for name conflicts, and the trap would be triggered earlier as well. Something like

```suggestion
          (
          TMP_DIR=$(mktemp -d)
          trap "rm --force --recursive $TMP_DIR" EXIT
          cp "${config.xdg.configHome}/gh/hosts.yml" $TMP_DIR/
          export GH_CONFIG_DIR=$TMP_DIR
          $DRY_RUN_CMD ${getExe cfg.package} help 2>&1 > $DRY_RUN_NULL
          cp $TMP_DIR/hosts.yml "${config.xdg.configHome}/gh/hosts.yml"
          )
```

should work, I think.

---

Tested and squashed the change in.